subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c

```diff
@@ -1081,41 +1174,59 @@ int azure_iot_hub_disconnect(void)
 {
 	int err;
 
-	if (connection_state != STATE_CONNECTED) {
-		LOG_ERR("Azure IoT Hub is not connected");
-		return -EACCES;
+	if (!connection_state_verify(STATE_CONNECTED)) {
+		LOG_WRN("Azure IoT Hub is not connected");
+		return -ENOTCONN;
 	}
 
-	connection_state = STATE_DISCONNECTING;
+	connection_state_set(STATE_DISCONNECTING);
 
 	err = mqtt_disconnect(&client);
 	if (err) {
 		LOG_ERR("Failed to disconnect MQTT client, error: %d", err);
+		connection_state_set(STATE_INIT);
 		return err;
 	}
 
-	connection_state = STATE_INIT;
+	/* The MQTT library only propogates the MQTT_DISCONNECT event
+	 * if the call to mqtt_disconnect() is successful. In that case the
+	 * setting of STATE_INIT is carried out in the mqtt_evt_handler.
+	 */
 
 	return 0;
 }
 
 int azure_iot_hub_connect(void)
 {
-	if (connection_state != STATE_INIT) {
-		LOG_ERR("Azure IoT Hub is not initialized");
-		return -EACCES;
+	int err;
+
+	if (connection_state_verify(STATE_CONNECTING)) {
```

Comment:
If connection is rejected on MQTT-level, it can be seen in the CONNACK event. I think we need to reset the state from connecting to init in `mqtt_evt_handler()` if that happens. Otherwise we're stuck in connecting state without actually attempting to connect

Reply:
Nice, Didn't see that one.